invesalius/gui/widgets/canvas_renderer.py

```diff
@@ -16,26 +15,51 @@
 #    COMERCIALIZACAO ou de ADEQUACAO A QUALQUER PROPOSITO EM
 #    PARTICULAR. Consulte a Licenca Publica Geral GNU para obter mais
 #    detalhes.
-#--------------------------------------------------------------------------
+# --------------------------------------------------------------------------
 
 import sys
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Dict,
+    List,
+    Literal,
+    Optional,
+    Tuple,
+    Union,
+    overload,
+)
+from weakref import WeakMethod
 
 import numpy as np
 import wx
+from typing_extensions import Self
+from vtkmodules.vtkRenderingCore import vtkActor2D, vtkCoordinate, vtkImageMapper
 
-try:
-    from weakref import WeakMethod
-except ImportError:
-    from weakrefmethod import WeakMethod
+from invesalius.data import converters
 
-from vtkmodules.vtkRenderingCore import vtkImageMapper, vtkActor2D, vtkCoordinate
+if TYPE_CHECKING:
+    from vtkmodules.vtkRenderingCore import vtkRenderer
 
-from invesalius.data import converters
+    from invesalius.data.viewer_slice import Viewer as sliceViewer
+    from invesalius.data.viewer_volume import Viewer as volumeViewer
+    from invesalius.gui.bitmap_preview_panel import SingleImagePreview as bitmapSingleImagePreview
+    from invesalius.gui.dicom_preview_panel import SingleImagePreview as dicomSingleImagePreview
 
 
 class CanvasEvent:
-    def __init__(self, event_name, root_event_obj, pos, viewer, renderer,
-                 control_down=False, alt_down=False, shift_down=False):
+    def __init__(
+        self,
+        event_name: str,
+        root_event_obj,
+        pos,
+        viewer,
```

Comment:
type Union[sliceViewer, volumeViewer, bitmapSingleImagePreview, dicomSingleImagePreview],

Reply:
I noticed that `get_vtk_mouse_position` is not implemented for both `SingleImagePreview` classes even though they are considered as viewers. Do I just create both functions that raise `NotImplementedError`?

Reply:
Good idea!